crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -225,6 +226,109 @@ impl IndexerReader {
+        let object_version_info: Option<StoredObjectVersion> =
+            run_query_async!(&pool, move |conn| {
+                let version_condition: Box<
+                    dyn BoxableExpression<objects_version::table, diesel::pg::Pg, SqlType = Bool>,
+                > = if before_version {
+                    Box::new(objects_version::object_version.lt(object_version_num))
+                } else {
+                    Box::new(objects_version::object_version.eq(object_version_num))
+                };
+
+                objects_version::dsl::objects_version
+                    .filter(objects_version::object_id.eq(object_id_bytes_clone))
+                    .filter(version_condition)
+                    .order_by(objects_version::object_version.desc())
+                    .limit(1)
+                    .first::<StoredObjectVersion>(conn)
+                    .optional()
+            })?;
+
+        let Some(object_version_info) = object_version_info else {
+            // Check if the object ever existed.
+            let pool = self.get_pool();
+            let object_id_bytes_clone = object_id_bytes.clone();
+            let latest_existing_version: Option<i64> = run_query_async!(&pool, move |conn| {
+                objects_version::dsl::objects_version
+                    .filter(objects_version::object_id.eq(object_id_bytes_clone))
+                    .order_by(objects_version::object_version.desc())
+                    .select(objects_version::object_version)
+                    .limit(1)
+                    .first::<i64>(conn)
+                    .optional()
+            })?;
+
+            return match latest_existing_version {
+                Some(latest_version) => {
+                    if object_version_num > latest_version {
+                        Ok(PastObjectRead::VersionTooHigh {
+                            object_id,
+                            asked_version: object_version,
+                            latest_version: SequenceNumber::from(latest_version as u64),
+                        })
+                    } else {
+                        Ok(PastObjectRead::VersionNotFound(object_id, object_version))
+                    }
+                }
+                None => Ok(PastObjectRead::ObjectNotExists(object_id)),
+            };
+        };
+
+        // Query objects_history for the object with the requested version.
+        let pool = self.get_pool();
+        let object_id_bytes_clone = object_id_bytes.clone();
+        let history_object: Option<StoredHistoryObject> = run_query_async!(&pool, move |conn| {
+            // Match directly on the primary key.
+            let query = objects_history::dsl::objects_history
+                .filter(
+                    objects_history::checkpoint_sequence_number
+                        .eq(object_version_info.cp_sequence_number),
+                )
+                .filter(objects_history::object_id.eq(object_id_bytes_clone))
```

Comment:
Nit: avoid the `_clone` suffix
```suggestion
        let object_id_bytes = object_id.to_vec();
        let history_object: Option<StoredHistoryObject> = run_query_async!(&pool, move |conn| {
            // Match directly on the primary key.
            let query = objects_history::dsl::objects_history
                .filter(
                    objects_history::checkpoint_sequence_number
                        .eq(object_version_info.cp_sequence_number),
                )
                .filter(objects_history::object_id.eq(&object_id_bytes))
```